File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int min = nums[0];
        int max = nums[0];
        int res = nums[0];
        for(int i=1; i<nums.length; i++) {
            int minEl = Math.min(nums[i], Math.min(nums[i]*min, nums[i] * max));
            int maxEl = Math.max(nums[i], Math.max(nums[i]*min, nums[i] * max));
            min = minEl;
            max = maxEl;
            res = Math.max(res, max);
        }
        return res;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    let max = 1
    let min = 1
    let res = Math.max(...nums)
    for(let i=0; i<nums.length; i++) {
        const num = nums[i]
        const tempMax = max*num
        max = Math.max(num, max*num, min*num)
        console.log(max, " : ", tempMax)
        min = Math.min(num, tempMax, min*num)
        res = Math.max(res, min, max)
    }
    return res
};